src/server/game/BattlePets/BattlePetMgr.cpp

```diff
@@ -677,6 +678,72 @@ void BattlePetMgr::ChangeBattlePetQuality(ObjectGuid guid, BattlePetBreedQuality
     // _owner->GetPlayer()->SetCurrentBattlePetBreedQuality(qualityValue);
 }
 
+void BattlePetMgr::GrantBattlePetExperience(ObjectGuid guid, uint16 xp)
+{
+    if (!HasJournalLock())
+        return;
+
+    BattlePet* pet = GetPet(guid);
+    if (!pet)
+        return;
+
+    if (xp <= 0)
+        return;
+
+    if (BattlePetSpeciesEntry const* battlePetSpecies = sBattlePetSpeciesStore.LookupEntry(pet->PacketInfo.Species))
+        if (battlePetSpecies->GetFlags().HasFlag(BattlePetSpeciesFlags::CantBattle))
+            return;
+
+    uint16 level = pet->PacketInfo.Level;
+    if (level >= MAX_BATTLE_PET_LEVEL)
+        return;
+
+    GtBattlePetXPEntry const* xpEntry = sBattlePetXPGameTable.GetRow(level);
+    if (!xpEntry)
+        return;
+
+    Player* player = _owner->GetPlayer();
+    uint16 nextLevelXp = uint16(GetBattlePetXPPerLevel(xpEntry));
+    // TODO: Calculate XP granted with auras of type SPELL_AURA_MOD_BATTLE_PET_XP_PCT (doesn't affect the experience granted by SPELL_EFFECT_GRANT_BATTLEPET_EXPERIENCE)
+    xp += pet->PacketInfo.Exp;
+
+    while (xp >= nextLevelXp && level < MAX_BATTLE_PET_LEVEL)
+    {
+        xp -= nextLevelXp;
+
+        xpEntry = sBattlePetXPGameTable.GetRow(++level);
+        if (!xpEntry)
+            return;
+
+        nextLevelXp = uint16(GetBattlePetXPPerLevel(xpEntry));
+
+        player->UpdateCriteria(CriteriaType::BattlePetReachLevel, pet->PacketInfo.Species, level);
```

Comment:
Why level is a miscValue2?

Reply:
In order to satisfy certain ModifierTreeType that use miscValue1 as a battle pet species (e.g. https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Achievements/CriteriaHandler.cpp#L2077)